otherlibs/systhreads/thread.ml

```diff
@@ -34,15 +34,33 @@ external exit_stub : unit -> unit = "caml_thread_exit"
 
 let[@inline never] check_memprof_cb () = ref ()
 
+let default_uncaught_exception_handler = thread_uncaught_exception
+
+let uncaught_exception_handler = ref default_uncaught_exception_handler
+
+let set_uncaught_exception_handler fn = uncaught_exception_handler := fn
+
 let create fn arg =
   thread_new
     (fun () ->
       try
         fn arg;
         ignore (Sys.opaque_identity (check_memprof_cb ()))
       with exn ->
-             flush stdout; flush stderr;
-             thread_uncaught_exception exn)
+        let raw_backtrace = Printexc.get_raw_backtrace () in
+        flush stdout; flush stderr;
+        try
+          !uncaught_exception_handler exn
+        with exn' ->
+          Printf.eprintf
+            "Fatal error in thread %d : %s\n"
```

Comment:
Nit: the other exception processing doesn't have a space before the colon (same in the next message)

Reply:
I wondered at first glance as to why the message was different from the "Thread %d killed on uncaught exception %s\n" on the C side, and the more I look at it, the more I think it _should_ be consistent.

The language (for good or ill!) of `Printexc.default_exception_handler` and `Printexc.handle_uncaught_exception'` uses "Fatal" because the _program_ is actually terminating. The errors here are only terminal for the thread, IIUC?

Reply:
Indeed these errors are only terminal for the thread.
I will stick to a more consistent language (akin to the messages found in the default C stub):
- `Thread %d killed on uncaught exception: %s` for the previously caught exception
- `Uncaught exception handler raised in thread %d: %s` for the handler's backtrace

Does this sound alright?